### ISSUE TITLE
Buttons for roles

### ISSUE THREAD

**Issue by tenhobi** [fit-ctu-discord/honza-botner#171]
- [ ] implement a simple prototype and make public test on main server

~~I would like a command feature, where mods can add or edit components (buttons to be specific) using a test message in some specific format or maybe JSON would be better.

We store these commands for message posting end editing, co we would do the same thing for this.

Take a look at how message add and edit works. Can this be replicated for adding and editing buttons?  

---

Will work on button setup and remove 
command. How should I handle config values for verification messages? Should I have simple command like `button setup verification` and it will automagically read config values and add buttons or should I remove that from config and do smth like `button setup verification %url%`?

_Originally posted by @stepech in https://github.com/fit-ctu-discord/honza-botner/pull/170#discussion_r711929547_~~

**Comment by tenhobi** [comment]
This probably needs some more thinking.

**Comment by stepech** [comment]
Might be autogenerated by bot.
When binding new role to fresh message, bot would add two buttons - `add role` and `remove role`. When tapping `add role`, bot would recognize what roles he can assign to user from that message and display only those the user doesn't have yet. He would do the same for `remove roles` but display only those the user has.
When binding new role to message with existing buttons, it would do nothing, just expand the selection menu after tapping one of those buttons.
Come with more ideas 😄

**Comment by ostorc** [comment]
The big difference between message editing and buttons editing is that, you have to have some support for the buttons action in the app and that is hardly doable from Discord. In my mind each button (or group of buttons) is somehow unique and shouldn't require any modding except first creation, which can be done directly from app.

**Comment by tenhobi** [comment]
Yep, we realized something like that in DMs. We will have to create some solid concept later. :)

**Comment by tenhobi** [comment]
We discussed buttons for roles today with @stepech and we find a quite nice solution.

The #bi-roles etc. channel would have a few buttons, we can split it in multiple messages to make it more clear:

- A - buttons that assign a user one specific role like `role for all courses and specializations`, and `role for szz`

// the rest of the buttons would not add ONE role, but it would send **EPHEMERAL** message to the user with generated select **select menu** filled with roles

- B - all available courses
- C - 6 buttons (that send an ephemeral message containing a select menu) with common courses for a given semester + 1 button for all common courses
- D - buttons for all specializations
- E - buttons for all departments

All these options in each select menu would have preselected those courses, that the user already has a role for.

---

### Scenario A -- course from Program's common courses + 1st semester course

I wanna add course BI-LA1, I have a few options:

- I can click a button that assigns me all BI courses (option A)
- I can click a button that sends me an ephemeral message that shows me all courses where I search it in a select menu (option B)
- I can click a button that sends me an ephemeral message that shows me a select menu with courses in 1st semester and select BI-LA1 (option C)
- // option D cannot be used here
- I can select a button for a given department that sends me an ephemeral message and select the course from a select menu (option E)

### Scenario B -- optional course

I wanna add course BI-VR1, I have a few options:

- I can click a button that assigns me all BI courses (option A)
- I can click a button that sends me an ephemeral message that shows me all courses where I search it in a select menu (option B)
- // option C cannot be used here
- // option D cannot be used here
- I can select a button for a given department that sends me an ephemeral message and select the course from a select menu (option E)

**Comment by ostorc** [comment]
Have you also considered how to manage the roles, the binding and so?

**Comment by tenhobi** [comment]
From binding with scheme `emoji, roleId` we would need to go to (probably some another dataset):

- `name` (BI-LA1)
- `description` (name of the course)
- `roleId` (Discord role ID)
- `selectMenuId` (not sure if needed or if we will manage to use only one ID for courses role granting)
- `specialization`
- `department`
- `semester`

**Comment by stepech** [comment]
Discord recently announced they have plans for special select menus for roles. Should come in 1st/2nd quarter of 2022. Current system works quite well... Let's wait?
